sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/InMemoryFileIndex.scala

```diff
@@ -94,10 +97,23 @@ class InMemoryFileIndex(
     val files = listLeafFiles(rootPaths)
     cachedLeafFiles =
       new mutable.LinkedHashMap[Path, FileStatus]() ++= files.map(f => f.getPath -> f)
-    cachedLeafDirToChildrenFiles = files.toArray.groupBy(_.getPath.getParent)
+    cachedLeafDirToChildrenFiles =
+      if (readPartitionWithSubdirectoryEnabled) {
+        files.toArray.groupBy(file => getRootPathsLeafDir(file.getPath.getParent))
```

Comment:
This will result in the inability to infer partitions for non catatlog table. Specifically, `"SPARK-20236: dynamic partition overwrite without catalog table"` will fail!